Update: Use elements mechanism for link color instead of a CSS variable

jorgefilipecosta opened the pull request:
This PR updates the link color mechanism to use elements instead of CSS variables.

The attribute is persisted in a paragraph is the same as in theme.json (inside elements). This is the first time we use an elements object at the block level. So we are introducing a new mechanism.
The shape looks like this:
```
<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#db0b15"}}}}} -->
```

The mechanism on the client-side is generic, e.g., it would work for any style and any element. The mechanism on the server is specified for link color for now. We can make it generic, but it requires exposing a method from lib/class-wp-theme-json.php that compiles the styles. To avoid having this PR change the public API of class-wp-theme-json.php, I removed the generic approach and kept things simpler and specific for link color so we can merge this PR faster.


## How has this been tested?

I went to the site editor global styles panel and set a global link color and a link color for the heading block.
I verified both link colors were applied correctly on the front and backend.
I went to the post editor and set a red link color for a paragraph. I verified the link color was applied correctly on the editor and frontend.

oandregal commented on lib/global-styles.php:
```diff
@@ -49,7 +49,6 @@ function gutenberg_experimental_global_styles_get_stylesheet( $tree, $type = 'al
  */
 function gutenberg_experimental_global_styles_enqueue_assets() {
 	if (
-		! get_theme_support( 'experimental-link-color' ) && // link color support needs the presets CSS variables regardless of the presence of theme.json file.
```
I'm thinking of this use case:

- a theme doesn't have a `theme.json` but it opts-in to link color via `add_theme_support('experimental-link-color')`
- the UI control will be available to the user, as a result, the block can have a `style.color.link` attribute which will be rendered by the element support mechanism

At this point, if the user selected one of the presets, none will be available. It looks like we still need to do these checks.

---

For further thoughts (not this PR): what's a good path to deprecate the existing `add_theme_support` mechanism?

jorgefilipecosta replied:
Good point I reverted the change.

oandregal replied:
ok, I've tested that this works in the front-end of TT1 as well. It doesn't work on the editor for for a different reason.

oandregal commented on lib/block-supports/elements.php:
```diff
@@ -0,0 +1,60 @@
+		$link_color      = "var(--wp--preset--color--$link_color_name)";
+	}
+
+	$style      = implode(
```
Can we have a single line for this (no tabs, no CRs, etc)? Fewer bytes and same readability.

jorgefilipecosta replied:
updated 👍

oandregal commented on lib/block-supports/elements.php:
```diff
@@ -0,0 +1,60 @@
+		"\n",
+		array(
+			'<style>',
+			".$class_name a{",
```
Do you think it's important that instead of `a` here we get the element selector from the existing [elements array](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json.php#L252)? I'm leaning towards using it.

jorgefilipecosta replied:
The elements array is not exposed is private information from that class in /lib/class-wp-theme-json.php.
I don't think we should expose the elements array. It is not something very useful to be part of the public API and elements were just merged this week. If we want to expose something I think we should expose a function that given a styles object compiles it to CSS (compute_style_properties) and in that case, we don't need to expose anything. But The public API is not yet decided, I would prefer if for now we don't expose anything and just use the "a" and then we expose some function to compile styles when our API is ready.

jorgefilipecosta commented:
Hi @nosolosw, thank you for the review your feedback was addressed. Meanwhile, this PR also got more complex because we were not dealing with some cases. The layout and duotone hooks assume there is always a class attribute present. In this case, we can not make that assumption (e.g: a simple paragraph). So I added code to deal with these cases where our code does not only needs to add a class but also needs to add the attribute.

oandregal commented on lib/block-supports/elements.php:
```diff
@@ -0,0 +1,63 @@
+	if ( null === $link_color ) {
+		return $block_content;
+	}
+	$class_name = 'wp-block-elements-container-' . uniqid();
```
These are the current classes:

- layout => `.wp-container-ID`
- duotone => `.wp-duotone-filter-ID`

For elements, how about `.wp-elements-ID`?

jorgefilipecosta replied:
Updated 👍

oandregal commented on lib/block-supports/elements.php:
```diff
@@ -0,0 +1,63 @@
+
+	$style = "<style>.$class_name a{color: $link_color;}</style>\n";
+
+	// Like the layout hook this assumes the hook only applies to blocks with a single wrapper.
```
What are examples of blocks that don't have a single top-level wrapper?

jorgefilipecosta replied:
I don't think we have a block without a single wrapper in core, but developers can create them.

oandregal commented on lib/block-supports/elements.php:
```diff
@@ -0,0 +1,63 @@
+ */
+
+/**
+ * Render out the duotone stylesheet and SVG.
```
```suggestion
 * Render the elements stylesheet.
```

oandregal commented on lib/block-supports/elements.php:
```diff
@@ -0,0 +1,63 @@
+	// Like the layout hook this assumes the hook only applies to blocks with a single wrapper.
+	// Retrieve the opening tag of the first HTML element.
+	$html_element_matches;
+	preg_match( '/<[^>]+>/', $block_content, $html_element_matches, PREG_OFFSET_CAPTURE );
```
I'd love to have a test for this function, given the use of this regexp and the below wrangling of replacing. Test cases:

- node without any attr
- node with a class attr
- node with an attr that is not a class

oandregal replied:
FWIW, tested in a variety of scenarios and it seems to work.

jorgefilipecosta replied:
I included some unit tests.

oandregal commented on lib/class-wp-theme-json.php:
```diff
@@ -836,53 +833,7 @@ private function get_block_styles( $style_nodes, $setting_nodes ) {
 			$selector     = $metadata['selector'];
 			$declarations = self::compute_style_properties( $node );
-
-			$is_link_element = self::is_link_element( $metadata['selector'] );
```
We can remove more bytes here that we no longer need (`is_link_element`, `without_link_selector`) :hocho: :hocho: :hocho:

jorgefilipecosta replied:
Done 👍

oandregal commented:
In `experimental-default-theme.json` we no longer need this bit:

```json
	"styles": {
		"elements": {
			"link": {
				"color": {
					"text": "#00E"
				}
			}
		}
	}
```

oandregal commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -75,6 +98,24 @@ export function getInlineStyles( styles = {} ) {
 
+function compileElementsStyles( selector, elements = {} ) {
+	return map( elements, ( styles, element ) => {
+		const linearStyles = getInlineStyles( styles );
```
How about `elementStyles` instead of `linearStyles`?

oandregal commented on packages/edit-site/src/components/editor/global-styles-renderer.js:
```diff
@@ -116,8 +116,8 @@ function flattenTree( input = {}, prefix, token ) {
 		STYLE_PROPERTY,
-		( declarations, { value, valueGlobal, properties }, key ) => {
-			const pathToValue = valueGlobal ?? value;
+		( declarations, { value, properties }, key ) => {
```
We can remove more bytes from this :hocho: :hocho: :hocho: (remove the logic that processed keys starting with `--`). We also need to prevent processing the `linkColor` property (properties whose value starts by `elements`).

jorgefilipecosta replied:
Fixed 👍

oandregal commented on packages/blocks/src/api/constants.js:
```diff
@@ -46,6 +41,10 @@ export const __EXPERIMENTAL_STYLE_PROPERTY = {
 		value: [ 'color', 'text' ],
 		support: [ 'color' ],
 	},
+	linkColor: {
```
How about we call this `colorLink` so it's alphabetized and consecutive to `color`?

oandregal replied:
Nah, forget about this. We also have `backgroundColor`, etc. It's not worth the effort to change the naming schema at this point.

oandregal commented on packages/block-editor/src/hooks/color.js:
```diff
@@ -303,17 +303,15 @@ export function ColorEdit( props ) {
+			colorObject?.slug
+				? `var:preset|color|${ colorObject.slug }`
+				: value,
+			clone
```
I'm confused about how this works and [the docs](https://lodash.com/docs/4.17.15#setWith) aren't of much help. Is there an alternative way to rewrite this? Wouldn't this work (pseudo-code)?

```js
props.setAttributes(
  style: {
    ...existingStyle,
   elements.link.color.text: <preset-or-raw-value>
  }
)
```

jorgefilipecosta replied:
Exactly what I'm doing is implementing this pseudocode. Basically, I want an immutable version of lodash set (because I don't want to mutate attributes), and as far as I know, a simple way to implement an immutable set is with setWith. To make the code easier to understand I extracted an immutableSet function.

oandregal commented:
I've done the first pass at this: the direction is good! Excited to have this in.

Some thoughts:

- It looks like some tests need updating.
- Do we want to migrate the existing blocks to the new way (remove the existing inline styles, migrate the block attribute, etc)?
- What are your thoughts on a path forward to remove the `experimental-link-color` theme support further down the road (not for this PR)?

jorgefilipecosta commented:
Hi @nosolosw, your feedback was applied, unit tests were added/updated.

> Do we want to migrate the existing blocks to the new way (remove the existing inline styles, migrate the block attribute, etc)?

I don't think we should migrate them, it will add more complexity and link color was and still is experimental. If some plugin or platform was relying on link color I guess the migration code should be added there.

> What are your thoughts on a path forward to remove the experimental-link-color theme support further down the road (not for this PR)?

I think will need this flag to enable preset variable outputs on themes without theme.json.

jorgefilipecosta committed:
Update: Use elements mechanism for link color instead of a CSS variable

Co-authored-by: André <nosolosw@users.noreply.github.com> (+3 squashed commits)
Squashed commits:
[994e5e3beb] Add link color as an element mechanism.
[e4a72bcb00] Remove existing link color
[7fcd57c9a6] Fix nav placeholder colors and height. (#31875)

oandregal commented:
@jorgefilipecosta looking at the CI report, `WP_Theme_JSON_Test::test_get_stylesheet` seems a legit failure.

oandregal commented on lib/block-supports/elements.php:
```diff
@@ -0,0 +1,65 @@
+	}
+	$link_color_declaration = esc_html( safecss_filter_attr( "color: $link_color" ) );
+
+	$style = "<style>.$class_name a{" . $link_color_declaration . ";}</style>\n";
```
```suggestion
	$style = "<style>.$class_name a{" . $link_color_declaration . " !important;}</style>\n";
```

This suggestion fixes this use case:

- TT1-blocks or TT1 theme.
- Add a paragraph with text and link colors applied.

The expected result is that both colors are applied but the link is not. This is because TT1-blocks has this rule `p.has-text-color a { color: inherit; }` with higher specificity than the class we add. While this is a theme issue that should be updated once this PR lands, we also need to cover for this sort of issues with themes. The way we address this in preset classes is by adding `!important` to the declaration (user styles are to be used above theme styles).

jorgefilipecosta replied:
I think we should probably fix the issue in the theme, using important for link color causes inconsistencies e.g: why are we using important for link color but not for other styles?
I guess we can temporarily include the important but we should plan on fixing the core themes and removing the important.

oandregal replied:
> I think we should probably fix the issue in the theme, using important for link color causes inconsistencies e.g: why are we using important for link color but not for other styles?

We actually do for other presets as well. See https://github.com/WordPress/gutenberg/pull/29533 for details and rationale.

oandregal commented:
I've added a paragraph with background, foreground, and link colors from the preset list. This is the serialized content:

```html
<!-- wp:paragraph {"style":{
  "elements":{"link":{"color":{"text":"var:preset|color|red"}}}},
  "backgroundColor":"gray",
  "textColor":"white"
} -->
<p class="has-white-color has-gray-background-color has-text-color has-background">
  paragraph with background and foreground and <a href="http://wordpress.org">link</a>
</p>
<!-- /wp:paragraph -->
```

Note how the `textColor` and `backgroundColor` only store the slug but the `linkColor` stores `var:preset|color|red`. Can we make `linkColor` work like the rest and only store the slug?

See related conversation at https://github.com/WordPress/gutenberg/pull/31762#issuecomment-839930411 with mobile folks. I'm thinking we want to get rid of the `var:preset|color|red` format, we don't need that now. That removal needs to be done in a follow-up PR as it also depends on refactoring the font-family to make it work like any other style property https://github.com/WordPress/gutenberg/pull/31910 In any case, for the concerns of this PR, we shouldn't use it if possible.

---

Edit: looked at [how we serialize the link color](https://github.com/WordPress/gutenberg/pull/31524/files#diff-b85df651c33208862897162a87698bad76f15535f6432b97b022a72e99230edeR31). Unlike `backgroundColor` and `textColor`, the `linkColor` block attribute can have either presets or raw values. I understand we use this format just to mark when the link color is a preset value. We could use a simpler mark such as `slug|value`, but the current format is also fine by me. I'm not opinionated on this. We can look independently to remove it from a few places.

oandregal commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -202,6 +243,45 @@ export const withBlockControls = createHigherOrderComponent(
+		if ( ! elements ) {
+			return <BlockListBlock { ...props } />;
+		}
+		const blockElementsContainerIdentifier = `wp-block-elements-container-${ useInstanceId(
```
I think we should use [the same as in the server](https://github.com/WordPress/gutenberg/pull/31524/files#diff-b85df651c33208862897162a87698bad76f15535f6432b97b022a72e99230edeR29): `wp-elements-ID`

oandregal commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -202,6 +243,45 @@ export const withBlockControls = createHigherOrderComponent(
+		const blockElementsContainerIdentifier = `wp-block-elements-container-${ useInstanceId(
+			BlockListBlock
+		) }`;
+		const styles = compileElementsStyles(
```
The style generated for the link color needs to include `!important`, same logic as in [the server](https://github.com/WordPress/gutenberg/pull/31524/files#r634252402).

Test case:

- Use the TT1 theme.
- Add a link color to a paragraph.

The expected result is that it works, but it doesn't because the theme styles have higher specificity `.editor-styles-wrapper .wp-block a { color: var(--wp--style--color--link, var(--global--color-primary)); }`.

oandregal commented on packages/block-editor/src/hooks/color.js:
```diff
@@ -2,7 +2,7 @@
+import { isObject, setWith, clone } from 'lodash';
 
 /**
  * WordPress dependencies
```
The `addSaveProps` function needs its logic updated for assigning the `has-link-color` class. At the moment it checks for `style.color.link`.

jorgefilipecosta replied:
Updated. Right now the server-side hook does not apply this class. I think this class is not useful at all and we should probably remove it in a follow-up PR.

oandregal commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -31,6 +43,16 @@ const styleSupportKeys = [
 	SPACING_SUPPORT_KEY,
 ];
 
+export const ELEMENTS = {
```
Can we instead consolidate the existing `ELEMENTS` in `utils.js` and this one in a common place both the hook and the global styles rendered can access? It looks like `packages/blocks/src/api/constants.js` could be a good option, given we already define there the `__EXPERIMENTAL_STYLE_PROPERTY`.

oandregal commented:
I'm testing with existing content, to see how this behaves. This is what I've tried:

Using the `trunk` branch:

- added a paragraph with a link
- use a preset as the link color 
- publish
- check the serialized content and it is:

```html
<!-- wp:paragraph {"style":{"color":{"link":"var:preset|color|red"}}} -->
<p class="has-link-color" style="--wp--style--color--link:var(--wp--preset--color--red)">
Paragraph with <a href="http://wordpress.org">link</a>.
</p>
<!-- /wp:paragraph -->
```

Switch to this branch:

- load the post editor
- check the serialized content and it is (note the `p` wrapped in a `p`):

```html
<!-- wp:paragraph {"style":{"color":{"link":"var:preset|color|red"}}} -->
<p class="has-link-color">
<p class="has-link-color" style="--wp--style--color--link:var(--wp--preset--color--red)">
Paragraph with <a href="http://wordpress.org">link</a>.
</p>
</p>
<!-- /wp:paragraph -->
```

- select a preset color for link and update
- in the front-end the output is:

```html
<p class="wp-elements-60a3b78a854cd has-link-color"></p>
<p class="has-link-color" style="--wp--style--color--link:var(--wp--preset--color--red)">Paragraph with <a href="http://wordpress.org">link</a>.</p>
<p></p>
<style>.wp-elements-60a3b78a854cd a{color: var(--wp--preset--color--red);}</style>
```

- when I reload the post editor I see the block marked as invalid and the serialized has became:

```html
<!-- wp:paragraph {"style":{"color":{"link":"var:preset|color|red"},"elements":{"link":{"color":{"text":"var:preset|color|red"}}}}} -->
<p class="has-link-color">
<p class="has-link-color" style="--wp--style--color--link:var(--wp--preset--color--red)">
Paragraph with <a href="http://wordpress.org">link</a>.
</p>
</p>
<!-- /wp:paragraph -->
```

I think we have a couple of issues here:

- Can you reproduce this weird behavior on your end?
- If we don't migrate/clean block data for blocks that have existing link color data, we're going to have some conflicts (themes that still use the `--wp--style--color--link` that is inline, etc).

oandregal commented:
Ok, I've done a second pass at this. The major issues I've found are:

- Conflicts with themes (both TT1 and TT1-blocks). Proposed fixes for the [server](https://github.com/WordPress/gutenberg/pull/31524#discussion_r634300495) & [site editor](https://github.com/WordPress/gutenberg/pull/31524#discussion_r634252402).
- Existing blocks with link color applied ([see](https://github.com/WordPress/gutenberg/pull/31524#issuecomment-843145911)).

jorgefilipecosta commented:
> Note how the textColor and backgroundColor only store the slug but the linkColor stores var:preset|color|red. Can we make linkColor work like the rest and only store the slug?

> See related conversation at #31762 (comment) with mobile folks. I'm thinking we want to get rid of the var:preset|color|red format, we don't need that now. That removal needs to be done in a follow-up PR as it also depends on refactoring the font-family to make it work like any other style property #31910 In any case, for the concerns of this PR, we shouldn't use it if possible.


Hi @nosolosw, on the backgroundColor and textColor, we have two attributes one for named colors and one just for values so we can differentiate between color value "red" and present color "red". I think the system we have for link color is much much better. We use a single attribute. The shape of the style attribute is exactly the same as the shape of theme.json. When using a different shape people will be confused: Why on theme.json to set a paragraph link color I use style.elements.link.color for presets and color values but when creating a pattern to style a paragraph I use style.elements.link.color for color values but If I want to use a preset I need to use a linkColor attribute that is not available on theme.json?
This system also opens the possibility to reference variables other than preset variables.
In the textColor and backgroundColor case, the shape difference may be justified because of the class usage, but in this case, there is no class usage so there is no reason to have any difference in shape between what happens on the style attribute and what happens on theme.json.
Regarding the format var:preset|color|red instead of var(--wp--preset--color--red) it is not because of kses it is because the variable format when serialized as json looks like this var(\u002d\u002dwp\u002d\u002dpreset\u002d\u002dcolor\u002d\u002dred), so we created an easier to read format but is converted to variable format.

jorgefilipecosta commented:
Hi @nosolosw I applied your suggestions to be compatible with current theme styles and I mitigated the back compatibility issue (we are not breaking the markup of the paragraph anymore).

oandregal commented:
Ok, now this works in the tests I've done:

- Tested post editor with TT1 and TT1-blocks: paragraph with link; paragraph with link, foreground, and background colors; and a bunch of others (heading, site title, etc).
- Tested site editor with TT1-blocks.
- Existing content with link color doesn't break.

However, I'm a bit concerned about the resulting markup of existing content. Steps for testing:

- In the `trunk` branch, add a paragraph and set a link color from the presets.
- Switch to this branch.
- The paragraph doesn't have the color, so I applied it again.
- This is the result:

```html
<!-- wp:paragraph {"style":{"color":{"link":"var:preset|color|red"},"elements":{"link":{"color":{"text":"var:preset|color|red"}}}}} -->
<p class="has-link-color" style="--wp--style--color--link:var(--wp--preset--color--red)">
Paragraph with <a href="http://wordpress.org">link</a>.
</p>
<!-- /wp:paragraph -->
```

Note how the legacy mechanism for block attributes is still there (`style.color.link`) as well as the legacy custom property as inline styles (`--wp--style--color--link:var(--wp--preset--color--red)`). My concern is both that this can cause bugs and that we should migrate existing content.

Do you think we can ship this as it is and do the migration in a follow-up? Do you think we could modify this PR to use the existing block attribute `style.color.link` and migrate to the new `style.elements.link.color.text` in a follow-up?

oandregal commented:
These are the blocks that have support for link color and, hence, would need to be migrated: column, columns, group, heading, media-text, paragraph, post-author, post-comments, post-comments-form, post-comments-link, post-date, post-excerpt, post-terms, post-title, query-pagination-next, query-pagination-previous, site-title, template-part, term-description, verse.

I'm particularly concerned with the inline styles, as that could cause styling issues with themes that use the custom property. Perhaps unlikely for the majority of cases given our use of `!important` but still can happen. Having those unused block attributes isn't a big deal, although I think we should clean those up as well ― ideally, we'd reuse the existing data to style the link with the new mechanism.

Given the time constraints to get this in the RC, a compromise we could do is to ship this PR as it is and backport those migrations individually later ― if that's something folks agree with or makes sense in general.

oandregal commented:
Ran [this](https://github.com/WordPress/gutenberg/pull/31524#issuecomment-844043057) and [this](https://github.com/WordPress/gutenberg/pull/31524#issuecomment-844197139) by some folks. Given this has been an experimental feature not shipped to core, there's going to be a small number of users that need those migrations. The consensus is that this is an important PR to have in today's RC and that we can fix bugs later should they happen. Going to [disagree and commit](https://en.wikipedia.org/wiki/Disagree_and_commit) with this one.
